tests/zfs-tests/tests/functional/fault/auto_offline_001_pos.ksh

```diff
@@ -0,0 +1,196 @@
+
+log_onexit cleanup
+
+filedev1="$TEST_BASE_DIR/file-vdev-1"
+filedev2="$TEST_BASE_DIR/file-vdev-2"
+filedev3="$TEST_BASE_DIR/file-vdev-3"
+sparedev="$TEST_BASE_DIR/file-vdev-spare"
+removedev=$(get_debug_device)
+
+typeset poolconfs=("mirror $filedev1 $removedev"
+    "raidz $filedev1 $removedev"
+    "raidz2 $filedev1 $filedev2 $removedev"
+    "raidz3 $filedev1 $filedev2 $filedev3 $removedev"
+    "$filedev1 cache $removedev"
+    "mirror $filedev1 $filedev2 cache $removedev"
+    "raidz $filedev1 $filedev2 $filedev3 cache $removedev"
+)
+
+truncate -s $SPA_MINDEVSIZE $filedev1
+truncate -s $SPA_MINDEVSIZE $filedev2
+truncate -s $SPA_MINDEVSIZE $filedev3
+truncate -s $SPA_MINDEVSIZE $sparedev
+
+for conf in "${poolconfs[@]}"
+do
+	# 1. Create a pool
+	log_must zpool create -f $TESTPOOL $conf
+	# NOTE: the following 'udevadm settle' helps avoid race conditions
+	# caused by mixing creation/removal events from partitioning the disk
+	# (zpool create) and events from physically removing it (remove_disk).
+	udevadm settle
+
+	# 2. Simulate physical removal of one device
+	remove_disk $removedev
+
+	# 3. Verify the device is offlined
+	log_must wait_vdev_state $TESTPOOL $removedev "OFFLINE"
+
+	# 4. Reattach the device
+	insert_disk $removedev
+
+	# 5. Verify the device is onlined
+	log_must wait_vdev_state $TESTPOOL $removedev "ONLINE"
+
+	# cleanup
+	destroy_pool $TESTPOOL
+	log_must zpool labelclear "/dev/${removedev}1"
+	zed_events_drain
+done
+
+# 6. Repeat the same tests with a spare device: zed will use the spare to handle
+#    the removed data device
+for conf in "${poolconfs[@]}"
+do
+	# 1. Create a pool with a spare
+	log_must zpool create -f $TESTPOOL $conf
+	log_must zpool add $TESTPOOL spare $sparedev
+	# NOTE: the following 'udevadm settle' helps avoid race conditions
+	# caused by mixing creation/removal events from partitioning the disk
+	# (zpool create) and events from physically removing it (remove_disk).
+	udevadm settle
+
+	# 3. Simulate physical removal of one device
+	remove_disk $removedev
+
+	# 4. Verify the device is handled by the spare unless is a l2arc disk
+	# which can only be offlined
+	if [[ $(echo "$conf" | grep -c 'cache') -eq 0 ]]; then
+		log_must wait_hotspare_state $TESTPOOL $sparedev "INUSE"
+	else
+		log_must wait_vdev_state $TESTPOOL $removedev "OFFLINE"
+	fi
+
+	# 5. Reattach the device
+	insert_disk $removedev
+
+	# 6. Verify the device is onlined
+	log_must wait_vdev_state $TESTPOOL $removedev "ONLINE"
+
+	# cleanup
+	destroy_pool $TESTPOOL
+	log_must zpool labelclear "/dev/${removedev}1"
+	zed_events_drain
+done
+
+# 7. Repeat the same tests again with a faulted spare device: zed should offline
+#    the removed data device if no spare is available
+for conf in "${poolconfs[@]}"
+do
+	# 1. Create a pool with a spare
+	log_must zpool create -f $TESTPOOL $conf
+	# XXX: remove this, debugging centos-6 builder failure which cannot be reproduced on my environment
```

Comment:
I think this is a race condition, the test seems to work on my CentOS6 builder at home; i regrettably spent more time debugging race conditions in the test case than writing code for this PR.

I am going to try a different approach and use `file_wait $ZED_LOG` instead of `zed_events_drain` since the latter does not wait for events coming from udev.